pkg/lightning/errormanager/errormanager.go

```diff
@@ -686,75 +802,95 @@ func (em *ErrorManager) ReplaceConflictKeys(
 					if err != nil {
 						return errors.Trace(err)
 					}
-					// calculate the new mustKeepKvPairs corresponding to the new rawKey
-					// find out all the KV pairs that are contained in the data KV
-					mustKeepKvPairs = encoder.SessionCtx.TakeKvPairs()
-				}
-			}
 
-			// if the latest value of rawKey equals to rawValue, that means this data KV is maintained in downstream TiDB
-			// if not, that means this data KV has been deleted due to overwritten index KV
-			if bytes.Equal(rawValue, latestValue) {
-				continue
-			}
-
-			handle, err := tablecodec.DecodeRowKey(rawKey)
-			if err != nil {
-				return errors.Trace(err)
-			}
-			decodedData, _, err := tables.DecodeRawRowData(encoder.SessionCtx,
-				tbl.Meta(), handle, tbl.Cols(), rawValue)
-			if err != nil {
-				return errors.Trace(err)
-			}
-			if !tbl.Meta().HasClusteredIndex() {
-				// for nonclustered PK, need to append handle to decodedData for AddRecord
-				decodedData = append(decodedData, types.NewIntDatum(handle.IntValue()))
-			}
-			_, err = encoder.Table.AddRecord(encoder.SessionCtx.GetTableCtx(), decodedData)
-			if err != nil {
-				return errors.Trace(err)
-			}
+					// find out all the KV pairs that are contained in the data KV
+					kvPairs := encoder.SessionCtx.TakeKvPairs()
+					for _, kvPair := range kvPairs.Pairs {
+						em.logger.Debug("got encoded KV",
+							logutil.Key("key", kvPair.Key),
+							zap.Binary("value", kvPair.Val))
+						kvLatestValue, err := fnGetLatest(dataGCtx, kvPair.Key)
```

Comment:
there is another potential point, we can use "batch get" to retrieve the result.